File: trie.cpp
#include <iostream>
using namespace std;
#include <string>
#include <vector>
#include <map>

map<string, string> dictname;
map<string, string> dictnum;

struct NameNode{
    NameNode *links[26];
    bool flag = false;
};

class NameTrie {
private: NameNode *root;
public:
    NameTrie(){
        root = new NameNode(); 
    }
    
    int insert(string name) {
        NameNode *node = root;
        for(int i=0; i<name.size(); i++){
            if( node->links[ name[i] - 'a'] == NULL ){
                node->links[ name[i] - 'a'] = new NameNode();
            }
            node = node->links[ name[i] - 'a'];
        }
        if(node->flag==true){
            // cout<<"Already present"<<endl;
            return -1;
        }
        node->flag = true;
        return 1;
    }

    bool search(string name) {
        NameNode *node = root;
        for(int i=0; i<name.size(); i++){
            cout<< name[i]<<" "<<node->links[ name[i] - 'a' ]<<endl;
            if( node->links[ name[i] - 'a' ] == NULL ) return false;
            node = node->links[ name[i] - 'a'];
        }
        if( node->flag == true ) return true;
        return false;
    }

    bool startsWith(string prefix) {
        NameNode *node = root;
        for(int i=0; i<prefix.size(); i++){
            cout<< prefix[i]<<" "<<node->links[ prefix[i] - 'a' ]<<endl;
            if( node->links[ prefix[i] - 'a'] == NULL ) return false;
            node = node->links[ prefix[i] - 'a'];
        }
        return true;
    }

    bool isLastNode(NameNode* node){
        for(int i=0; i<26; i++){
            if(node->links[i]!=NULL) return false;
        }
        return true;
    }

    void rec(NameNode* node, string name, vector<string>& store){
        if(node->flag == true){
            // cout<<"found flag as true "<<name<<endl;
            store.push_back(name);
        }
        for(int i=0; i<26; i++){
            // cout<<"looping for"<<i<<" "<<name<<endl;
            if(node->links[i]!=NULL){
                // cout<<"found "<<i<<" "<<name<<endl;
                name += (char)(i + 'a');
                rec(node->links[i], name, store);
                // cout<<"going back to "<<i<<" "<<name<<endl;
                name.erase(name.end()-1);
            }
        }
        return;
    }
    void print(){
        NameNode* node = root;
        vector<string> store;
        // cout<<"Start Printing "<<endl;
        rec(node, "", store);
        // To print name only
        // for(int i=0; i<store.size(); i++) cout<<store[i]<<endl;
        // To print name along with contact
        for(int i=0; i<store.size(); i++) cout<<store[i]<<" - "<<dictname[store[i]]<<endl;
    }
    int printpref(string pref){
        NameNode* node = root;
        for(int i=0; i<pref.size(); i++){
            if(node->links[ pref[i] - 'a' ] == NULL){
                // cout<<pref[i]<<" "<<node->links[pref[i]-'a']<<endl;
                // cout<< "not present" <<endl;
                return 0;
            }
            // cout<<pref[i]<<" "<<node->links[pref[i]-'a']<<" not NULL"<<endl;
            node = node->links[ pref[i] - 'a' ];
        }
        if(isLastNode(node)){
            cout<<pref<<endl;
            return -1;
        }
        vector<string> store;
        rec(node, pref, store);
        // To print name only
        // for(int i=0; i<store.size(); i++) cout<<store[i]<<endl;
        // To print name along with contact
        for(int i=0; i<store.size(); i++) cout<<store[i]<<" - "<<dictname[store[i]]<<endl;
        return 1;
    }
};

struct ContactNode{
    ContactNode* links[10];
    bool flag = false;
};

class ContactTrie {
private: ContactNode *root;
public:
    ContactTrie(){
        root = new ContactNode(); 
    }
    
    int insert(string number) {
        ContactNode *node = root;
        for(int i=0; i<number.size(); i++){
            if( node->links[ number[i] - '0'] == NULL ){
                node->links[ number[i] - '0'] = new ContactNode();
            }
            node = node->links[ number[i] - '0'];
        }
        if(node->flag==true){
            // cout<<"Already present"<<endl;
            return -1;
        }
        node->flag = true;
        return 1;
    }

    bool search(string number) {
        ContactNode *node = root;
        for(int i=0; i<number.size(); i++){
            cout<< number[i]<<" "<<node->links[ number[i] - '0' ]<<endl;
            if( node->links[ number[i] - '0' ] == NULL ) return false;
            node = node->links[ number[i] - '0'];
        }
        if( node->flag == true ) return true;
        return false;
    }

    bool startsWith(string prefix) {
        ContactNode *node = root;
        for(int i=0; i<prefix.size(); i++){
            cout<< prefix[i]<<" "<<node->links[ prefix[i] - '0' ]<<endl;
            if( node->links[ prefix[i] - '0'] == NULL ) return false;
            node = node->links[ prefix[i] - '0'];
        }
        return true;
    }

    bool isLastNode(ContactNode* node){
        for(int i=0; i<10; i++){
            if(node->links[i]!=NULL) return false;
        }
        return true;
    }

    void rec(ContactNode* node, string number, vector<string>& store){
        if(node->flag == true){
            // cout<<"found flag as true "<<name<<endl;
            store.push_back(number);
        }
        for(int i=0; i<10; i++){
            // cout<<"looping for"<<i<<" "<<name<<endl;
            if(node->links[i]!=NULL){
                // cout<<"found "<<i<<" "<<name<<endl;
                number += (char)(i + '0');
                rec(node->links[i], number, store);
                // cout<<"going back to "<<i<<" "<<name<<endl;
                number.erase(number.end()-1);
            }
        }
        return;
    }
    void print(){
        ContactNode* node = root;
        vector<string> store;
        // cout<<"Start Printing "<<endl;
        rec(node, "", store);
        // To print contact only
        // for(int i=0; i<store.size(); i++) cout<<store[i]<<endl;
        // To print contact along with name
        for(int i=0; i<store.size(); i++) cout<<store[i]<<" - "<<dictnum[store[i]]<<endl;
    }
    int printpref(string pref){
        ContactNode* node = root;
        for(int i=0; i<pref.size(); i++){
            if(node->links[ pref[i] - '0' ] == NULL){
                // cout<<pref[i]<<" "<<node->links[pref[i]-'a']<<endl;
                // cout<< "not present" <<endl;
                return 0;
            }
            // cout<<pref[i]<<" "<<node->links[pref[i]-'a']<<" not NULL"<<endl;
            node = node->links[ pref[i] - '0' ];
        }
        if(isLastNode(node)){
            cout<<pref<<endl;
            return -1;
        }
        vector<string> store;
        rec(node, pref, store);
        // To print contact only
        // for(int i=0; i<store.size(); i++) cout<<store[i]<<endl;
        // To print contact along with name
        for(int i=0; i<store.size(); i++) cout<<store[i]<<" - "<<dictnum[store[i]]<<endl;
        return 1;
    }
};

int main(){
    freopen("in.txt", "r", stdin);
    freopen("out.txt", "w", stdout);
    NameTrie nametrie;
    ContactTrie contacttrie;
    int t;
    cin>>t;
    while(t--){
        int i;
        cin>>i;
        while(i--){
            string name;
            string contact;
            cin>>name;
            cin>>contact;
            dictname[name] = contact;
            dictnum[contact] = name;
            int n = nametrie.insert(name);
            if(n==-1) cout<<"Name already present"<<endl;
            int m = contacttrie.insert(contact);
            if(m==-1) cout<<"Contact already present"<<endl;
        }
        // cout<<trie.search("abc");
        // cout<<trie.startsWith("ab");
        nametrie.print();
        cout<<endl;
        nametrie.printpref("sh");
        cout<<endl;
        nametrie.printpref("him");
        cout<<endl;
        contacttrie.print();
        cout<<endl;
        contacttrie.printpref("6367");
        cout<<endl;
        contacttrie.printpref("86194");
        cout<<endl;
    }
    return 0;
}

File: trie.js
const dictname = new Map()
const dictnum = new Map()

class NameNode {
  constructor() {
    this.links = new Array(26)
    this.flag = false
  }
}

class NameTrie {
  constructor() {
    this.root = new NameNode()
  }

  insert(name) {
    let node = this.root
    for (let i = 0; i < name.length; i++) {
      const index = name.charCodeAt(i) - 'a'.charCodeAt(0)
      if (node.links[index] === undefined) {
        node.links[index] = new NameNode()
      }
      node = node.links[index]
    }
    if (node.flag) {
      return -1
    }
    node.flag = true
    return 1
  }

  search(name) {
    let node = this.root
    for (let i = 0; i < name.length; i++) {
      const index = name.charCodeAt(i) - 'a'.charCodeAt(0)
      if (node.links[index] === undefined) {
        return false
      }
      node = node.links[index]
    }
    return node.flag
  }

  isLastNode(node) {
    for (let i = 0; i < 26; i++) {
      if (node.links[i] !== undefined) {
        return false
      }
    }
    return true
  }

  rec(node, name, store) {
    if (node.flag) {
      store.push(name)
    }
    for (let i = 0; i < 26; i++) {
      if (node.links[i] !== undefined) {
        const nextChar = String.fromCharCode(i + 'a'.charCodeAt(0))
        this.rec(node.links[i], name + nextChar, store)
      }
    }
  }

  print() {
    const node = this.root
    const store = []
    this.rec(node, '', store)
    console.log(store.length)
    for (let i = 0; i < store.length; i++) {
      console.log(store[i] + ': ' + dictname[store[i]])
    }
  }

  printPref(pref) {
    let node = this.root
    for (let i = 0; i < pref.length; i++) {
      const index = pref.charCodeAt(i) - 'a'.charCodeAt(0)
      if (node.links[index] === undefined) {
        console.log('No contact found with the given user input')
        return false
      }
      node = node.links[index]
    }
    if (this.isLastNode(node)) {
      console.log(pref + ':' + dictname[pref])
      return true
    }
    const store = []
    this.rec(node, pref, store)
    console.log(store.length)
    for (let i = 0; i < store.length; i++) {
      console.log(store[i] + ': ' + dictname[store[i]])
    }
    return true
  }
}

class ContactNode {
  constructor() {
    this.links = new Array(10)
    this.flag = false
  }
}

class ContactTrie {
  constructor() {
    this.root = new ContactNode()
  }

  insert(number) {
    let node = this.root
    for (let i = 0; i < number.length; i++) {
      const index = number.charCodeAt(i) - '0'.charCodeAt(0)
      if (node.links[index] === undefined) {
        node.links[index] = new ContactNode()
      }
      node = node.links[index]
    }
    if (node.flag) {
      return -1
    }
    node.flag = true
    return 1
  }

  search(number) {
    let node = this.root
    for (let i = 0; i < number.length; i++) {
      const index = number.charCodeAt(i) - '0'.charCodeAt(0)
      if (node.links[index] === undefined) {
        return false
      }
      node = node.links[index]
    }
    return node.flag
  }

  isLastNode(node) {
    for (let i = 0; i < 10; i++) {
      if (node.links[i] !== undefined) {
        return false
      }
    }
    return true
  }

  rec(node, number, store) {
    if (node.flag) {
      store.push(number)
    }
    for (let i = 0; i < 10; i++) {
      if (node.links[i] !== undefined) {
        const nextDigit = String.fromCharCode(i + '0'.charCodeAt(0))
        this.rec(node.links[i], number + nextDigit, store)
      }
    }
  }

  print() {
    const node = this.root
    const store = []
    this.rec(node, '', store)
    console.log(store.length)
    for (let i = 0; i < store.length; i++) {
      console.log(dictnum[store[i]] + ': ' + store[i])
    }
  }

  printPref(pref) {
    let node = this.root
    for (let i = 0; i < pref.length; i++) {
      const index = pref.charCodeAt(i) - '0'.charCodeAt(0)
      if (node.links[index] === undefined) {
        console.log('No contact found with the given user input')
        return false
      }
      node = node.links[index]
    }
    if (this.isLastNode(node)) {
      console.log(dictnum[pref] + ':' + pref)
      return true
    }
    const store = []
    this.rec(node, pref, store)
    console.log(store.length)
    for (let i = 0; i < store.length; i++) {
      console.log(dictnum[store[i]] + ': ' + store[i])
    }
    return true
  }
}

const names = new NameTrie();
const numbers = new ContactTrie();

function add() {
  const inpname = document.getElementById('addname');
  const inpnumber = document.getElementById('addcontact');
  const name = inpname.value;
  const number = inpnumber.value;

  dictname[name] = number;
  dictnum[number] = name;
  const nameInsertResult = names.insert(name);
  const numInsertResult = numbers.insert(number);
  if (nameInsertResult === 1 && numInsertResult === 1) {
    console.log('Contact added successfully.');
  } else {
    console.log('Contact already exists.');
  }
}
function searchbyname() {
  const inpsearchName = document.getElementById('getbyname');
  const searchName = inpsearchName.value;
  if (names.search(searchName)) {
    console.log('Name found: ' + searchName + ': ' + dictname[searchName]);
  } else {
    console.log('Name not found.');
  }
}
function searchbycontact() {
  const inpsearchContact = document.getElementById('getbycontact');
  const searchContact = inpsearchContact.value;
  if (numbers.search(searchContact)) {
    console.log(
      'Contact found: ' + dictnum[searchContact] + ': ' + searchContact
    )
  } else {
    console.log('Contact not found.');
  }
}
function printall() {
  console.log('All Contacts:');
  names.print();
}
function printnameinit() {
  const inpsearchInitialName = document.getElementById('initname');
  const searchInitialName = inpsearchInitialName.value;
  names.printPref(searchInitialName);
}
function printcontactinit() {
  const inpsearchInitialContact = document.getElementById('initcontact');
  const searchInitialContact = inpsearchInitialContact.value;
  numbers.printPref(searchInitialContact);
}

function main() {
  var myButton1 = document.getElementById('mybtn1')
  var myButton2 = document.getElementById('mybtn2')
  var myButton3 = document.getElementById('mybtn3')
  var myButton4 = document.getElementById('mybtn4')
  var myButton5 = document.getElementById('mybtn5')
  var myButton6 = document.getElementById('mybtn6')

  myButton1.addEventListener('click', add);
  myButton2.addEventListener('click', searchbyname);
  myButton3.addEventListener('click', searchbycontact);
  myButton4.addEventListener('click', printall);
  myButton5.addEventListener('click', printnameinit);
  myButton6.addEventListener('click', printcontactinit);

}

main()
